### PR TITLE
fix(core): generic solution for issue #111 for all command groups

### PR DESCRIPTION
**Pull request by maximilien**
BFS exploration of all sub-commands from `rootCmd` and adds a `RunE` for all commands that are command groups with child sub-commands and without a `RunE`.

The added `RunE` will return the correct errors when the command is called with empty sub-command or with an unknown sub-command. It will also print the command help message first.

----
I did not add a test for this since the we need to decide on #185 and #214 and #134. Once we do I will add the gotest.tools or Ginkgo test for it.

### CODE REVIEW

**Comment by maximilien** [comment]
/assign @cppforlife

cc: @rhuss

File: pkg/kn/core/root.go [hunk 1]
```diff
@@ -71,11 +72,33 @@ Eventing: Manage event subscriptions and channels. Connect up event sources.`,
...
+// deal with errors when called with empty or unknown sub command
+func emptyAndUnknownSubcommands(cmd *cobra.Command) {
+	for _, childCmd := range cmd.Commands() {
+		if childCmd.HasSubCommands() && childCmd.RunE == nil {
+			childCmd.RunE = func(aCmd *cobra.Command, args []string) error {
+				aCmd.Help()
+				if len(args) == 0 {
+					return errors.New(fmt.Sprintf("please provide a valid sub-command for \"kn %s\"", aCmd.Name()))
+				} else {
+					return errors.New(fmt.Sprintf("unknown command \"%s\" for \"kn %s\"", args[0], aCmd.Name()))
```
Comment (navidshaikh) [inline pkg/kn/core/root.go]: *unkown sub-command

**Comment by rhuss** [comment]
>     3\. Can we add some tests please?

I think @maximilien wants just to wait with tests until with which testing setup we want to continue. So let's wait until tomorrows WG meeting.

File: pkg/kn/core/root.go (hunk 1 above)
Comment (maximilien) [reply pkg/kn/core/root.go]: Hmm, did not see the changes to the docs. I’ll try again tomorrow.

Yes, once we agree I will add the tests as we agree on the route for tests.

Comment (maximilien) [reply pkg/kn/core/root.go]: OK addressed this in latest push. Resolving. Re-open if I did not. Thx.

**Comment by maximilien** [comment]
OK @rhuss, I think I am convinved `gotests.tools` is better for this project. Added a test for `root.go` that also tests the functionality for issue #111 and I like the result. See latest update here.

cc: @navidshaikh and @cppforlife

**Comment by rhuss** [comment]
@maximilien I think you missed to check in the vendored files (use a `build.sh -u` to populate the vendor dir)

The PR looks good to me, let's merge it tomorrow if there are no objections against gotest.tools in today's WG call.

**Comment by maximilien** [comment]
> @maximilien I think you missed to check in the vendored files (use a `build.sh -u` to populate the vendor dir)
> 
> The PR looks good to me, let's merge it tomorrow if there are no objections against gotest.tools in today's WG call.

Yes, thanks @rhuss, one day I will remember to do this before pushing :) Converting all ginkgo tests now. Best.

**Comment by rhuss** [comment]
@maximilien cool, thanks ! 

Now that gotest.tools is in, we have to rebase our other PRs which introduce them (but should be hopefully trivial)

**Pull request merged**